### ISSUE TITLE
[Bug]: Incorrect memory reporting while running in lxc container

### ISSUE THREAD

**Issue by aldem** [netdata/netdata#14238]
### Bug description

Memory reporting is incorrect when agent is running inside of Proxmox LCX container under systemd and kernel 5.19 - instead of reporting LXC's memory it reports host's memory.

Actually this is not netdata's bug, the problem is that when option `ProtectSystem=full` is active, systemd mounts proc fs from the host - this might be a bug either in systemd or LXC related code in the Linux kernel, maybe even specific to only some Kernel versions (I didn't  research this deep yet).

### Expected behavior

Netdata should report container's RAM usage.

This could be achieved by adding additional option to `netdata.service` file:

`BindReadOnlyPaths=/proc:/proc:rbind`

### Steps to reproduce

1. Install netdata inside of LXC container running Proxmox under kernel 5.19
2. Run it and you will see host memory instead of container's.

### Installation method

kickstart.sh

### System info

```shell
Linux pbs 5.19.17-1-pve #1 SMP PREEMPT_DYNAMIC PVE 5.19.17-1 (Mon, 14 Nov 2022 20:25:12  x86_64 GNU/Linux
/etc/os-release:PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
/etc/os-release:NAME="Debian GNU/Linux"
/etc/os-release:VERSION_ID="11"
/etc/os-release:VERSION="11 (bullseye)"
/etc/os-release:VERSION_CODENAME=bullseye
/etc/os-release:ID=debian
```

### Netdata build info

```shell
Version: netdata v1.37.1
Configure options:  '--prefix=/opt/netdata/usr' '--sysconfdir=/opt/netdata/etc' '--localstatedir=/opt/netdata/var' '--libexecdir=/opt/netdata/usr/libexec' '--libdir=/opt/netdata/usr/lib' '--with-zlib' '--with-math' '--with-user=netdata' '--enable-cloud' '--without-bundled-protobuf' '--disable-dependency-tracking' 'CFLAGS=-static -O2 -I/openssl-static/include -pipe' 'LDFLAGS=-static -L/openssl-static/lib' 'PKG_CONFIG_PATH=/openssl-static/lib/pkgconfig'
Install type: kickstart-static
    Binary architecture: x86_64
Features:
    dbengine:                   YES
    Native HTTPS:               YES
    Netdata Cloud:              YES
    ACLK:                       YES
    TLS Host Verification:      YES
    Machine Learning:           YES
    Stream Compression:         YES
Libraries:
    protobuf:                YES (system)
    jemalloc:                NO
    JSON-C:                  YES
    libcap:                  NO
    libcrypto:               YES
    libm:                    YES
    tcalloc:                 NO
    zlib:                    YES
Plugins:
    apps:                    YES
    cgroup Network Tracking: YES
    CUPS:                    NO
    EBPF:                    YES
    IPMI:                    NO
    NFACCT:                  NO
    perf:                    YES
    slabinfo:                YES
    Xen:                     NO
    Xen VBD Error Tracking:  NO
Exporters:
    AWS Kinesis:             NO
    GCP PubSub:              NO
    MongoDB:                 NO
    Prometheus Remote Write: YES
Debug/Developer Features:
    Trace Allocations:       NO
```

**Comment by ilyam8** [comment]
Hi, @aldem. Thanks for opening the ticket!

> This could be achieved by
> BindReadOnlyPaths=/proc:/proc:rbind

I checked it and it is more complex than this. I assume it is not only memory, but the desired result is to monitor the container (from inside the container) in general. Using "BindReadOnlyPaths" solves the problem for memory (partially) and for CPUs.

I said "partially", because the memory is completely wrong if using ZFS (which is likely on Proxmox). We take into account ZFS ARC cache:

https://github.com/netdata/netdata/blob/b513888be389f92b2323d1bb3fdf55c22d4e4bad/collectors/proc.plugin/proc_meminfo.c#L161-L163

And regardless of the use of "BindReadOnlyPaths=/proc" 

- the kernel exposes hosts' ZFS metrics inside an LXC container (`/proc/spl/kstat/zfs/`). 
- also, `/proc/diskstats` returns hosts devices too.
- likely something else.

**Comment by aldem** [comment]
If I am not completely wrong, `/proc/diskstats` is anyway not virtualized, containers see all host devices in "default" configuration (with or without systemd magic) - thus I doubt we can do something here. I believe that ZFS is exposed due to same bug/feature - and I am not aware of any way to prevent or disable this.

However, there is a solution - if netdata detects (or explicitly signalled in configuration) that it is running inside of LXC container, it may omit diskstats and ZFS memory calculations just because inside of LXC this makes little sense. Unless, of course, someone wants to monitor host from inside of LXC (in docker?) - but then the default configuration (without using `BindReadOnlyPaths`) does the trick anyway.

Another option is to allow to explicitly specify during installation how to handle this case - this way installation script may create an overlay and use `BindReadOnlyPaths` if requested or (by default) installation inside LXC is detected.

From my point of view, proper memory metrics are quite important so we could disregard the rest (diskstats & ZFS in particular), for now at least.

**Comment by ilyam8** [comment]
>  it may omit diskstats and ZFS memory calculations

That is what `htop` does when [calculating memory usage](https://github.com/htop-dev/htop/blob/4d40680aa04370b06e9a2acd732f0be2ee6b6eba/linux/Platform.c#L362-L371)

**Comment by ilyam8** [comment]
Did some research. I am not sure, but I think `ProtectSystem=full` has nothing to do with the problem.

> If MountAPIVFS= is used, let's mount /sys, /proc, /dev and /run into the it, but only as a fallback if the user hasn't mounted something there already. These mounts are hence overridden by any other explicitly configured mounts.

[apivfs_table](https://github.com/systemd/systemd/blob/5228c58ebe8fb77d579f333263986a0798299ab7/src/core/namespace.c#L101-L108) (includes `/proc`).

[Add those mounts](https://github.com/systemd/systemd/blob/5228c58ebe8fb77d579f333263986a0798299ab7/src/core/namespace.c#L2303-L2307) from the table if [namespace_info_mount_apivfs()](https://github.com/systemd/systemd/blob/5228c58ebe8fb77d579f333263986a0798299ab7/src/core/namespace.c#L1643-L1657) which results to true **if any**

```c
ns_info->protect_control_groups ||
ns_info->protect_kernel_tunables ||
ns_info->protect_proc != PROTECT_PROC_DEFAULT ||
ns_info->proc_subset != PROC_SUBSET_ALL
```

Any of the following options results in mounting host's `/proc`: 

- `ProtectControlGroups`
- `ProtectKernelTunables`
- `ProtectProc` (any non-default value)
- `ProcSubset` (any non-default value)


In our case it is [ProtectControlGroups=on](https://github.com/netdata/netdata/blob/5980b18ecd0a83d91c783bca597da7fb4156ce19/system/netdata.service.in#L69).

**Comment by ilyam8** [comment]
I checks `/proc/mounts` w/ and w/o enabled `ProtectControlGroups`. This is the difference @aldem:

```go
lxcfs /proc/cpuinfo fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/diskstats fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/loadavg fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/meminfo fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/stat fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/swaps fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/uptime fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /sys/devices/system/cpu/online fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
```

^^ Those entries are missing **when enabled**. It explains why ZFS stats are exposed from inside the container.

**Comment by ilyam8** [comment]
Fixed in #14255

**Comment by ilyam8** [comment]
Just realized that we have a problem with loadavg after #14255. Load average virtualization is disabled by default in LXCFS (`-l, --enable-loadavg Enable loadavg virtualization`) - it means that all [loadavg alarms](https://github.com/netdata/netdata/blob/master/health/health.d/load.conf) will be active because the calculation is based on the number of processors. 🤷‍♂️ how to fix that properly.
